tests/MSBuildDeviceIntegration/Tests/InstallAndRunTests.cs

```diff
@@ -702,8 +702,30 @@ public void RunWithLLVMEnabled ()
 			else
 				AdbStartActivity ($"{proj.PackageName}/{proj.JavaPackageName}.MainActivity");
 
-			Assert.IsTrue (WaitForActivityToStart (proj.PackageName, "MainActivity",
-				Path.Combine (Root, builder.ProjectDirectory, "startup-logcat.log")));
+			var activityNamespace   = proj.PackageName;
+			var activityName        = "MainActivity";
+			var logcatFilePath      = Path.Combine (Root, builder.ProjectDirectory, "startup-logcat.log");
+			var failedToLoad        = new List<string> ();
+			bool appLaunched        = MonitorAdbLogcat ((line) => {
```

Comment:
@grendello, @dellis1972: how does this updated test look for verifying that `.so` files were all loaded successfully?

Reply:
It verifies by looking for error messages saying that a library **failed** to load, the next line

Reply:
@grendello, @dellis1972 : [my attempted unit test change failed](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6994251&view=ms.vss-test-web.build-test-results-tab&runId=61803194&resultId=100030&paneView=debug).  Is that because things aren't working properly, or because the test is looking for the wrong thing?

```
LLVM .so files not loaded:
11-21 23:05:25.104 10192 10192 I monodroid-assembly: Failed to load shared library '/data/app/com.xamarin.runwithinterpreterenabled-EBknjyBcFiejLCycXre3mw==/split_config.x86_64.apk!/lib/x86_64/UnnamedProject.dll.so'. dlopen failed: library "/data/app/com.xamarin.runwithinterpreterenabled-EBknjyBcFiejLCycXre3mw==/split_config.x86_64.apk!/lib/x86_64/UnnamedProject.dll.so" not found
11-21 23:05:25.105 10192 10192 I monodroid-assembly: Failed to load shared library 'UnnamedProject.dll.so'. dlopen failed: library "UnnamedProject.dll.so" not found
11-21 23:05:25.106 10192 10192 I monodroid-assembly: Failed to load shared library '/data/app/com.xamarin.runwithinterpreterenabled-EBknjyBcFiejLCycXre3mw==/split_config.x86_64.apk!/lib/x86_64/Mono.Android.dll.so'. dlopen failed: library "/data/app/com.xamarin.runwithinterpreterenabled-EBknjyBcFiejLCycXre3mw==/split_config.x86_64.apk!/lib/x86_64/Mono.Android.dll.so" not found
11-21 23:05:25.106 10192 10192 I monodroid-assembly: Failed to load shared library 'Mono.Android.dll.so'. dlopen failed: library "Mono.Android.dll.so" not found
11-21 23:05:25.107 10192 10192 I monodroid-assembly: Failed to load shared library '/data/app/com.xamarin.runwithinterpreterenabled-EBknjyBcFiejLCycXre3mw==/split_config.x86_64.apk!/lib/x86_64/Java.Interop.dll.so'. dlopen failed: library "/data/app/com.xamarin.runwithinterpreterenabled-EBknjyBcFiejLCycXre3mw==/split_config.x86_64.apk!/lib/x86_64/Java.Interop.dll.so" not found
11-21 23:05:25.108 10192 10192 I monodroid-assembly: Failed to load shared library 'Java.Interop.dll.so'. dlopen failed: library "Java.Interop.dll.so" not found
11-21 23:05:25.116 10192 10192 I monodroid-assembly: Failed to load shared library '/data/app/com.xamarin.runwithinterpreterenabled-EBknjyBcFiejLCycXre3mw==/split_config.x86_64.apk!/lib/x86_64/Mono.Android.Runtime.dll.so'. dlopen failed: library "/data/app/com.xamarin.runwithinterpreterenabled-EBknjyBcFiejLCycXre3mw==/split_config.x86_64.apk!/lib/x86_64/Mono.Android.Runtime.dll.so" not found
11-21 23:05:25.116 10192 10192 I monodroid-assembly: Failed to load shared library 'Mono.Android.Runtime.dll.so'. dlopen failed: library "Mono.Android.Runtime.dll.so" not found
11-21 23:05:25.122 10192 10192 I monodroid-assembly: Failed to load shared library '/data/app/com.xamarin.runwithinterpreterenabled-EBknjyBcFiejLCycXre3mw==/split_config.x86_64.apk!/lib/x86_64/System.Runtime.dll.so'. dlopen failed: library "/data/app/com.xamarin.runwithinterpreterenabled-EBknjyBcFiejLCycXre3mw==/split_config.x86_64.apk!/lib/x86_64/System.Runtime.dll.so" not found
11-21 23:05:25.122 10192 10192 I monodroid-assembly: Failed to load shared library 'System.Runtime.dll.so'. dlopen failed: library "System.Runtime.dll.so" not found
11-21 23:05:25.585 10192 10192 I monodroid-assembly: Failed to load shared library '/data/app/com.xamarin.runwithinterpreterenabled-EBknjyBcFiejLCycXre3mw==/split_config.x86_64.apk!/lib/x86_64/System.Console.dll.so'. dlopen failed: library "/data/app/com.xamarin.runwithinterpreterenabled-EBknjyBcFiejLCycXre3mw==/split_config.x86_64.apk!/lib/x86_64/System.Console.dll.so" not found
11-21 23:05:25.585 10192 10192 I monodroid-assembly: Failed to load shared library 'System.Console.dll.so'. dlopen failed: library "System.Console.dll.so" not found
Expected: 0
But was:  12
```

My guess is that `Failed to load shared library` *always* appears, because not all assemblies are AOT'd. Which seems odd, as `$(EnableLLVM)`=true is set; shouldn't that AOT *all* of them?

Also, those names look weird, e.g. `System.Runtime.dll.so` can't be found.  Shouldn't it be `libaot-System.Runtime.dll.so`?

Reply:
Lines ending with `not found` should be ignored, since it's perfectly fine that an AOT library wasn't packaged (e.g. if it was empty) or even created. However, testing for this properly would require finding a library that references one of the symbols in libc/libm and check if that library is loaded. However, I can't find any such library.

@jonathanpeppers originally [found](https://discord.com/channels/732297728826277939/732297837953679412/1032301219391209522) the problem in `libaot-Mono.Android.dll.so`:

```
so I was testing LLVM with the .NET 7 RC, and noticed it was really slow...
Looking at full logcat I get:
10-18 23:55:17.844 20389 20389 D monodroid-assembly: monodroid_dlopen: hash match found, DSO name is 'libaot-Mono.Android.dll.so'
10-18 23:55:17.844 20389 20389 I monodroid-assembly: Trying to load shared library '/data/app/~~4kGTzxWG6HKO7nzyg6ryBg==/com.companyname.hellomaui.old-h8Jlutcvf8Dzfpuuq1ouUA==/lib/arm64/libaot-Mono.Android.dll.so'
10-18 23:55:17.844 20389 20389 I monodroid-assembly: Failed to load shared library '/data/app/~~4kGTzxWG6HKO7nzyg6ryBg==/com.companyname.hellomaui.old-h8Jlutcvf8Dzfpuuq1ouUA==/lib/arm64/libaot-Mono.Android.dll.so'. dlopen failed: cannot locate symbol "memset" referenced by "/data/app/~~4kGTzxWG6HKO7nzyg6ryBg==/com.companyname.hellomaui.old-h8Jlutcvf8Dzfpuuq1ouUA==/lib/arm64/libaot-Mono.Android.dll.so"...

Do you have an idea of what went wrong there?
Basically none of the AOT images are loading
```

And indeed, when you look at symbols referenced by `libaot-Mono.Android.dll.so` from the apk (see the attached zip), you can see both the references **and** a `text` segment:

```
$ llvm-nm -CD libaot-Mono.Android.dll.so | head -10
0000000000048450 W gc.safepoint_poll
                 U memset
00000000000d89d0 D mono_aot_file_info
00000000000a9f80 T p_101_plt_Java_Net_Proxy_get_NoProxy_llvm
00000000000aa00c T p_108_plt_Android_Runtime_JNIEnv_FindClass_string_llvm
00000000000aa020 T p_109_plt_Android_Runtime_JNIEnv_GetMethodID_intptr_string_string_llvm
00000000000aa034 T p_110_plt_Android_Runtime_JNIEnv_DeleteGlobalRef_intptr_llvm
00000000000aa070 T p_113_plt_Android_Runtime_JNIEnv_CallObjectMethod_intptr_intptr_llvm
00000000000aa0d4 T p_118_plt_Java_Net_HttpURLConnection__ctor_intptr_Android_Runtime_JniHandleOwnership_llvm
00000000000aa0e8 T p_119_plt_Android_Runtime_XAPeerMembers__ctor_string_System_Type_llvm
```

However, `Mono.Android.dll.so` produced with XA/main has just the data `mono_aot_file_info` symbol:

```
$ llvm-nm -CD Mono.Android.dll.so | head -10
0000000000062840 D mono_aot_file_info
```

@jonathanpeppers's test was with one of the NET7 RC builds, so perhaps something has changed inbetween? Also with regard to library names?

[libraries.zip](https://github.com/xamarin/xamarin-android/files/10065293/libraries.zip)

Reply:
Is the test passing because `$(AndroidNdkDirectory)` is set and valid?

When I hit the original problem, I was building on Azure DevOps with an NDK in a non-standard location -- would be the same as not having an NDK installed at all.

I solved the problem by doing:

```yaml
-p:AndroidNdkDirectory=$(ANDROID_NDK_HOME)
```